File: src/application/use_cases/inventory.py
"""Inventory-focused application use cases."""

from __future__ import annotations

from pathlib import Path

from src.application.ports import InventoryCatalog, InventoryRefresher
from src.domain.models import ClusterTarget


def list_cluster_targets(
    inventory_path: Path,
    catalog: InventoryCatalog,
) -> list[ClusterTarget]:
    return catalog.list_targets(inventory_path)


def find_target_by_context_name(
    context_name: str,
    inventory_path: Path,
    catalog: InventoryCatalog,
) -> ClusterTarget | None:
    for target in list_cluster_targets(inventory_path, catalog):
        if target.context_name == context_name:
            return target
    return None


def deduplicate_context_names(context_names: list[str]) -> list[str]:
    unique_contexts: list[str] = []
    seen: set[str] = set()
    for context_name in context_names:
        if context_name in seen:
            continue
        seen.add(context_name)
        unique_contexts.append(context_name)
    return unique_contexts


def select_targets_by_context_name(
    context_names: list[str],
    inventory_path: Path,
    catalog: InventoryCatalog,
) -> tuple[list[ClusterTarget], list[str]]:
    available_targets = list_cluster_targets(inventory_path, catalog)
    by_context = {target.context_name: target for target in available_targets}
    selected: list[ClusterTarget] = []
    missing: list[str] = []

    for context_name in context_names:
        target = by_context.get(context_name)
        if target is None:
            missing.append(context_name)
            continue
        selected.append(target)

    return selected, missing


def refresh_inventory_if_possible(
    inventory_path: Path,
    refresher: InventoryRefresher,
) -> tuple[bool, str] | None:
    if not inventory_path.exists():
        return None
    return refresher.refresh(inventory_path)

File: internal/application/usecases/inventory.go
package usecases

import (
	"os"

	"github.com/systemframe/k3ctx/internal/application"
	"github.com/systemframe/k3ctx/internal/domain"
)

// ListClusterTargets returns all targets from the inventory catalog.
func ListClusterTargets(inventoryPath string, catalog application.InventoryCatalog) ([]domain.ClusterTarget, error) {
	return catalog.ListTargets(inventoryPath)
}

// FindTargetByContextName returns the first target matching contextName, or nil.
func FindTargetByContextName(contextName, inventoryPath string, catalog application.InventoryCatalog) (*domain.ClusterTarget, error) {
	targets, err := ListClusterTargets(inventoryPath, catalog)
	if err != nil {
		return nil, err
	}
	for _, t := range targets {
		if t.ContextName() == contextName {
			cp := t
			return &cp, nil
		}
	}
	return nil, nil
}

// DeduplicateContextNames returns context names in order with duplicates removed.
func DeduplicateContextNames(names []string) []string {
	seen := map[string]bool{}
	out := []string{}
	for _, n := range names {
		if !seen[n] {
			seen[n] = true
			out = append(out, n)
		}
	}
	return out
}

// SelectTargetsByContextName resolves context names to ClusterTargets.
// Returns (selected, missing) — missing contains names not found in catalog.
func SelectTargetsByContextName(contextNames []string, inventoryPath string, catalog application.InventoryCatalog) ([]domain.ClusterTarget, []string, error) {
	available, err := ListClusterTargets(inventoryPath, catalog)
	if err != nil {
		return nil, nil, err
	}
	byContext := map[string]domain.ClusterTarget{}
	for _, t := range available {
		byContext[t.ContextName()] = t
	}

	var selected []domain.ClusterTarget
	var missing []string
	for _, name := range contextNames {
		if t, ok := byContext[name]; ok {
			selected = append(selected, t)
		} else {
			missing = append(missing, name)
		}
	}
	return selected, missing, nil
}

// RefreshInventoryIfPossible calls refresher when the path exists or is empty.
// An empty path means the active catalog (e.g. NetBird) manages its own source;
// the refresher is called directly without a filesystem check.
// Returns nil when the path is non-empty but does not exist on disk.
func RefreshInventoryIfPossible(inventoryPath string, refresher application.InventoryRefresher) *[2]any {
	if inventoryPath != "" {
		if _, err := os.Stat(inventoryPath); err != nil {
			return nil
		}
	}
	ok, msg := refresher.Refresh(inventoryPath)
	result := [2]any{ok, msg}
	return &result
}
